lib/ProductOpener/Display.pm

```diff
@@ -5467,30 +5464,12 @@ sub display_pagination($$$$) {
 					my $link;
 
 					if (defined $current_link) {
-						$link = $current_link;
-						#check if groupby_tag is used
-						if (defined $request_ref->{groupby_tagtype}) {
-							if (("/" . $request_ref->{groupby_tagtype}) ne $current_link) {
-								$link = $current_link . "/" . $request_ref->{groupby_tagtype}
-							}
-						}
-						if ($i > 1) {
-							$link .= "/$i";
-						}
-						if ($link eq '') {
-							$link  = "/";
-						}
-						if (defined $request_ref->{sort_by}) {
-							$link .= "?sort_by=" . $request_ref->{sort_by};
-						}
-					}
```

Comment:
Won't it be a regression to remove that ?  I would think that previously we were going through this block ? (because $current_link was defined in most cases). Although I agree this is not compatible with current_link in search.pl (which did already have parameter, while this block clearly applies to an url without parameters).

Reply:
I think this correspond to pages like https://world.openfoodfacts.net/brand/ferrero

Reply:
Hi @dipietroR , we do not have tests for the display_pagination() function yet, so you have to test your changes manually.

I think the first thing you could do is to add a comment in the POD format above the function to explain what it does (you can see examples of that for other functions that already have documentation above them).

display_pagination() generates a list of links to the different pages that correspond to a query. 
e.g. if you go to the bottom of https://world.openfoodfacts.org/
you will see:

![image](https://user-images.githubusercontent.com/8158668/171354023-d54fb325-1f34-4276-a622-9a0177fa1409.png)

Notice that the links have page numbers at the end like this: https://world.openfoodfacts.org/4

If you do a search query: https://world.openfoodfacts.org/cgi/search.pl?search_terms=coke&search_simple=1&action=process

Then you get page numbers, but the links are currently broken:
https://world.openfoodfacts.org/cgi/search.pl?action=process&search_terms=coke&sort_by=unique_scans_n&page_size=24/3?sort_by=unique_scans_n

They should be like this instead:

https://world.openfoodfacts.org/cgi/search.pl?action=process&search_terms=coke&sort_by=unique_scans_n&page_size=24&page=3

That was the intent of the code that you removed: if the link was a search query, then page numbers are to be passed as an extra parameter : &page=3.  If it's a facet, like https://world.openfoodfacts.org/label/organic , then we just add it like this : https://world.openfoodfacts.org/label/organic/3

To test it locally, you have very few products, so you won't get pagination unless you make an empty query:
http://world.openfoodfacts.localhost/cgi/search.pl?search_terms=&search_simple=1&action=process

The fact that it is currently broken in production today clearly shows that we should add some tests to catch this.

Could you try to add a test in display.t to test the result of the function display_pagination() with different types of $current_link and more than 1 page?

Reply:
I tried setting up 109 of the products in the test environment with the Organic label, though I couldn't get it to show the pagination that is shown on the production page (https://world.openfoodfacts.org/label/organic), Rather it only shows 100, then I can reach the second page by manually adding a /2 or ?page=2 at which point I see the remaining 9 products. Not sure why this is happening, but it is all happening on the main branch, rather than my branch. I'm pulling the full DB to test with that, but it is going to take awhile. Looks like it should be done in a few hours, so hopefully I can have a look with that tomorrow.

I can look into making a test for the display_pagination() function, though to verify, that should be part of a separate PR, correct?

Reply:
> I can look into making a test for the display_pagination() function, though to verify, that should be part of a separate PR, correct?

If you can it's a good thing to make separate PRs if things are independent, so that we can merge things incrementally without having to wait for every part to be perfect.

Reply:
Neat! no_cache=1 was just what I needed.

After applying no_cache=1 once. I don't seem to need to run it again, at which point on this PR I can go to:
http://world.openfoodfacts.localhost/label/organic?page=2&page_size=10
And the link to the first page will show:
http://world.openfoodfacts.localhost/label/organic?page=1&page_size=10
which indeed will go to the first page when clicked.

So far as I understand this PR is working, and fixes the bug that it, inadvertently, uncovered. Is there anything else that should be done on this before merging it?

Reply:
@dipietroR : there are 2 types of pages where we display a pagination:
- what we call "facets", with urls like /label/organic
- results of search queries with urls like /cgi/search.pl?...  or like https://world.openfoodfacts.org/search?labels_tags_fr=bio 

For the facets, we want the URLs of the different pages to use /[page number] at the end.
e.g. if you go to https://world.openfoodfacts.org/label/organic in production, the links in the pagination are to https://world.openfoodfacts.org/label/organic/2 etc.

This was done by the code removed above, we need to add back this logic. One way could be to check if $current_link contains a question mark, in which case we use the format &page=[page number], otherwise we add /[page number] when the page number is greater than 1.

Reply:
Also before merging the PR, it would be great to:

1. document the function in the POD format, to explain what it does, and the difference for the different page types (e.g. a summary of the info in this ticket).

2. add a few unit tests in display.t to test those different cases